### ISSUE TITLE
`forge build --extra-output ...` behaves a little unexpectedly

### ISSUE THREAD

**Issue by ehildenb** [foundry-rs/foundry#2309]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (7941935 2022-07-11T00:09:05.617669537Z)

### What command(s) is the bug in?

`forge build --extra-output evm.methodIdentifiers`

### Operating System

Linux

### Describe the bug

If you tell the `solc` compiler to include the field `evm.methodIdentifiers` and `evm.deployedBytecode.object` in it's output via the standard JSON interface (using this code here: https://github.com/runtimeverification/evm-semantics/blob/master/kevm_pyk/src/kevm_pyk/solc_to_k.py#L52), then I get a structure that looks like this:

```
{ ...
  "evm": { ...
                "methodeIdentifiers": ...
                "deployedBytecode": { .... "object": {...} ... }
         }
}
```

But adding arguments `--extra-output evm.methodIdentifiers --extra-output evm.deployedBytecode.object` produces something like this:

```
{ ...
  "methodIdentifiers": ...
  "deployedBytecode": { ... "object": { ... } ... }
}
```

Which is like it has flattened the `"evm"` selector out of the JSON heirarchy. So this makes it so I have to have a solc specific code path and a foundry specific code path for dealing with build artifacts in the above linked code, which is minorly annoying. It would be nice if Foundry used the same output format as solc exactly.

**Comment by ehildenb** [comment]
Here is the delta I had to make to handle this difference in format (it's small, so this is probably not super high priority, just an annoyance): https://github.com/runtimeverification/evm-semantics/pull/1280/commits/f053cf8d93642af7529c4178481f27181aa3957f

**Comment by onbjerg** [comment]
You can use `forge build --build-info` which places a JSON file with the exact input and output to/from solc. The reason these two are flattened is because we use them regularly, so they are always present; additionally, the JSON files in `out` (other than in `build-info`) are artifacts and are intended to be under our control in some capacity.

Does that solve it for you?

**Comment by ehildenb** [comment]
No, it doesn't really solve it for me, but happy for you to label it as won't-fix.

I think the tool should strive to be as compatible with the direct output of solc as possible, there seems to be no reason to avoid just having those fields nested. That way other tools have an easier time integrating with Foundry if they already have solc integration. Here I went and adapted my solc integration to the foundry output, and then had to spend about 20 minutes hunting down this breakage. Then I had to confirm for myself that the different structure indeed did have all the data I needed.

The actual code change (linked above) is small, but I can imagine that less brave developers (not that I'm very brave) would have just stopped at "well, it doesn't work with my tool that reads solc output, might as well give up."

**Comment by onbjerg** [comment]
It's a hard line between 1:1 solc or 1:1 Hardhat in terms of what we put in our artifacts. Right now we're neither, but I'm unsure if it's a good idea for us to go 1:1 solc since it leaves out the option for us to add additional things to the artifacts. Will keep this open as a low priority (since it's breaking) and we can discuss when we have room for bigger breaking changes

**Comment by ehildenb** [comment]
Does hardhat not use solc too? If they do, they should strive to be conformant with solc as well. Hardhat seems higher in the stack than solc to me.

**Comment by onbjerg** [comment]
They don't, they also have their own custom artifact format

**Comment by ehildenb** [comment]
Hardhat does use `solc` AFAICT: https://hardhat.org/hardhat-runner/docs/guides/compile-contracts

Does seem they should not be touching the artifacts solc outputs, at least not simple silly modifications like this which make very little actual difference.

**Comment by onbjerg** [comment]
That's not true: https://hardhat.org/hardhat-runner/docs/advanced/artifacts

Their artifacts are shaped like:

```
{
  contractName, abi, bytecode, deployedBytecode, linkReferences, deployedLinkReferences
}
```

Contrast that to solcs output: https://docs.soliditylang.org/en/v0.8.14/using-the-compiler.html#output-description

**Comment by ehildenb** [comment]
Yes, their artifacts are different, and there isn't much actionable you can do at this point.

I do think Hardhat is making a mistake here by reshaping the artifacts instead of reusing `solc`'s shaping (afterall, it is calling `solc` under the hood), but I don't think you need to keep this open.

**Comment by zerosnacks** [comment]
Closing this as `wontfix` per authors request and there is no further actionable request
